aries_cloudagent/utils/http.py

```diff
@@ -96,9 +101,65 @@ async def fetch(
                     response: ClientResponse = await session.get(url, headers=headers)
                     if response.status < 200 or response.status >= 300:
                         raise ClientError(
-                            f"Bad response from server: {response.status}"
+                            f"Bad response from server: {response.status} - "
+                            f"{response.reason}"
                         )
                     return await (response.json() if json else response.text())
             except (ClientError, asyncio.TimeoutError) as e:
                 if attempt.final:
                     raise FetchError("Exceeded maximum fetch attempts") from e
+
+
+async def put(
+    url: str,
+    file_data: dict,
+    extra_data: dict,
+    *,
+    retry: bool = True,
+    max_attempts: int = 5,
+    interval: float = 1.0,
+    backoff: float = 0.25,
+    request_timeout: float = 10.0,
+    connector: BaseConnector = None,
+    session: ClientSession = None,
+    json: bool = False,
+):
+    """Fetch from an HTTP server with automatic retries and timeouts.
+
+    Args:
+        url: the address to fetch
+        file_data: dict with data key and path of file to upload
+        extra_data: further content to include in data to put
+        headers: an optional dict of headers to send
+        retry: flag to retry the fetch
+        max_attempts: the maximum number of attempts to make
+        interval: the interval between retries, in seconds
+        backoff: the backoff interval, in seconds
+        request_timeout: the HTTP request timeout, in seconds
+        connector: an optional existing BaseConnector
+        session: a shared ClientSession
+        json: flag to parse the result as JSON
+
+    """
+    (data_key, file_path) = [k for k in file_data.items()][0]
+    data = {**extra_data}
+    limit = max_attempts if retry else 1
+
+    if not session:
+        session = ClientSession(connector=connector, connector_owner=(not connector))
+    async with session:
+        async for attempt in RepeatSequence(limit, interval, backoff):
```

Comment:
limit, interval, backoff. See utils/http.py. Parameters interval and backoff shape the power series heuristic for retry times.

Reply:
FYI in the bc registries issuer we use the "backoff" library for this, for example:

https://github.com/bcgov/von-bc-registries-agent/blob/aries-refactoring/data-pipeline/bcreg/credssubmitter.py#L114